clients/redshift/redshift.go

```diff
@@ -91,7 +93,47 @@ WHERE
 }
 
 func (s *Store) Dedupe(fqTableName string) error {
-	return fmt.Errorf("dedupe is not yet implemented")
+	stagingTableName := fmt.Sprintf("%s_dedupe_staging_%.5d", constants.ArtiePrefix, rand.Intn(100_000))
+
+	query := fmt.Sprintf(`
+CREATE TABLE %s AS SELECT DISTINCT * FROM %s;
+DELETE FROM %s;
+INSERT INTO %s SELECT * FROM %s;
+DROP TABLE %s;`,
+		// CREATE TABLE
+		stagingTableName,
+		// AS SELECT DISTINCT * FROM
+		fqTableName,
+		// ; DELETE FROM
+		fqTableName,
+		// ; INSERT INTO
+		fqTableName,
+		// SELECT * FROM
+		stagingTableName,
+		// ; DROP TABLE
+		stagingTableName,
+		// ;
+	)
+
+	transaction, err := s.Begin()
+	if err != nil {
+		return fmt.Errorf("failed to start transaction: %w", err)
+	}
+	defer func() {
+		if err := transaction.Rollback(); err != nil {
```

Comment:
@nathan-artie Just reading through this code again, doesn't this mean that rollback will always run?

Reply:
We can do something similar to this: https://github.com/artie-labs/transfer/blob/a857a7fd9521bb14933270483279185444f81aa5/clients/mssql/staging.go#L39

Reply:
It will, but from the [guide I was following](https://go.dev/doc/database/execute-transactions#:~:text=Defer%20the%20transaction%E2%80%99s%20rollback.%20If%20the%20transaction%20succeeds%2C%20it%20will%20be%20committed%20before%20the%20function%20exits%2C%20making%20the%20deferred%20rollback%20call%20a%20no%2Dop.%20If%20the%20transaction%20fails%20it%20won%E2%80%99t%20be%20committed%2C%20meaning%20that%20the%20rollback%20will%20be%20called%20as%20the%20function%20exits.):
> Defer the transaction’s rollback. If the transaction succeeds, it will be committed before the function exits, making the deferred rollback call a no-op. If the transaction fails it won’t be committed, meaning that the rollback will be called as the function exits.

I can add an check though to make it explicit.